[Token] Auth - Token 도메인 서비스 개발

codingbaraGo opened the pull request:
<!-- 제목은 `[도메인] PR 제목`으로 작성해주세요. (ex) [User] 소셜 로그인 필터 리팩토링 -->
<!-- PR의 타입은 Label을 통해 나타내주세요. -->

### 💻 작업 내용
<!-- 진행한 작업 내용에 대해 작성해주세요. -->
- 권한, 인증과 관련된 토큰 도메인 서비스 개발
- Stateless, Rotation를 사용한 토큰 인증 로직 개발
- 토큰 발급, 교체, 검증, 비활성화 로직 개발

### ✨ 리뷰 포인트
<!-- (ex) query 가 너무 복잡한 것 같은데 이 위주로 봐주세요. -->
PR이 너무 크니 세세한 코드보다는 TokenDomainService 클래스의 public method 별 흐름을 살피며 Stateless & Rotation이 잘 적용되었는지 위주로 봐주세요.
인증/인가의 경우에는 한 객체(엔티티)가 중심이 되어 로직이 흘러가기 보다는, 서비스 로직 위주로 흘러가고, 데이터들은 단지 비교/검증을 위해 사용되기만 합니다.
이런 특성을 고려하여 엔티티가 아닌 도메인서비스 클래스를 root로 잡고 구현하였습니다.

### 📝 메모
<!-- PR 관련 전달사항이 있다면 이곳에 작성해주세요. -->
Stateless(Blacklist) & Rotation 기반 인증 도메인 로직입니다.
현재 JwtProvider까지 개발을 완료했고, 이제 Repository들과 presentation 계층만 개발하면 됩니다.

### 🎯 관련 이슈
<!-- pr이 merge 되면 이슈가 자동으로 close 되도록 합니다. 만약 자동 close 를 하지 않고 이슈만 링크한다면 closed 키워드를 삭제해주세요.-->
#57

codingbaraGo committed:
feat(token): TokenClaims VO 정의 (#57)
- 토큰의 발급/검증에 사용되는 모든 정보를 가지는 Claims 클래스 정의

codingbaraGo committed:
feat(token): Token VO 정의 (#57)
- 상위 계층에서 사용할 Token 클래스 정의
- Access 토큰과 Refresh 토큰을 보관하는 TokenPair VO 정의

codingbaraGo committed:
feat(token): Repository 인터페이스 정의 (#57)
- MemberId를 UUID로 매핑해주는 MemberUUIDRepository 인터페이스 정의
- Token의 블랙리스트 repository 인터페이스 정의

codingbaraGo committed:
feat(token): TokenDomainService에서 사용할 유틸리티 클래스/인터페이스 정의 (#57)
- TokenProvider 인터페이스 정의
- 토큰 ID 생성기 구현
- 시간정보 관리 인터페이스 정의

codingbaraGo committed:
feat(token): TTL Policy 클래스 개발 (#57)
- 각 토큰별 TTL 정책 클래스 개발

codingbaraGo committed:
feat(token): TokenDomainService 개발
- login 토큰 발급, rotation, 인증 토큰 발급에 대한 DTO 클래스 정의
- 토큰 발급, rotation, 검증, 블랙리스트 등록 도메인 로직 개발

codingbaraGo committed:
test(token): TokenClaims 테스트 추가
- TokenClaims의 isExpired 로직 테스트 추가

codingbaraGo committed:
test(token): 도메인 서비스 테스트 개발
- Role 별 TTL 차등 적용 테스트 개발
- 도메인 서비스 메서드 별 성공/실패 테스트 개발

vwx12345 commented on src/main/java/com/simpleboard/board/auth/domain/token/service/TokenTTLPolicy.java:
```diff
@@ -0,0 +1,78 @@
+    else if (purpose.equals(VerifyPurpose.PASSWORD)) return Duration.ofMinutes(passwordMinutes);
+    else if (purpose.equals(VerifyPurpose.NICKNAME)) return Duration.ofMinutes(nicknameMinutes);
+    return Duration.ZERO;
+  }
```
switch 문을 사용안하고 if-else 를 사용하신 이유가 있을까요?

codingbaraGo replied:
switch로 수정했습니다.

vwx12345 commented on src/main/java/com/simpleboard/board/auth/domain/token/vo/Token.java:
```diff
@@ -0,0 +1,15 @@
+ * @since 1.0
+ */
+@Builder
+public record Token(String raw, String tokenId, String subject, Instant expiredAt) {}
```
record와 @Builder 조합을 잘 활용해주셔서 객체 생성이 간결하고 명확해진 점이 인상적이었습니다.
다만, 현재 상태에서는 각 필드에 null 값이 들어가는 것을 막을 수 있는 별도의 방어 로직이 없어 보입니다

codingbaraGo replied:
Validator는 presentation 계층 구성 후 추가하도록 하겠습니다.

vwx12345 commented on src/main/java/com/simpleboard/board/auth/domain/token/service/TokenDomainService.java:
```diff
@@ -0,0 +1,195 @@
+        tokenProvider.issueToken(createAccessClaims(command.memberId(), command.role(), now));
+
+    TokenClaims refreshClaims = createRefreshClaims(command.memberId(), command.role(), now);
+    Token refreshToken = tokenProvider.issueToken(refreshClaims);
```
사소한 부분일 수 있지만, accessToken은 createAccessClaims()를 바로 전달하고, refreshToken은 refreshClaims에 한 번 할당한 뒤 전달한 이유가 있을까요? 일관성 측면에서 동일한 방식으로 처리해도 괜찮을 것 같아 보여 간단히 여쭤봅니다.

codingbaraGo replied:
코드를 수정하면서 놓쳤나보네요.
반영하겠습니다!

vwx12345 commented on src/main/java/com/simpleboard/board/auth/domain/token/service/TokenDomainService.java:
```diff
@@ -0,0 +1,195 @@
+    // Validate refresh token
+    if (oldClaims.isExpired(now)) throw new RefreshTokenExpiredException();
+    if (blacklistRepository.exists(oldClaims.tokenId()))
+      throw new RefreshTokenInvalidException(); // TODO: 이벤트 생성
```
주석이 어떤 의미인지 모르겠습니다.

codingbaraGo replied:
요청과 함께 넘겨받은 refresh token의 만료와 블랙리스트 등록 여부를 체크하는 부분인데 어떤부분이 이해가지 않으신가요?
말 그대로 refresh token을 valitate하는 부분입니다.

vwx12345 replied:
'// TODO: 이벤트 생성' 부분 말하는 것이었습니다.

codingbaraGo replied:
아하
주석에 TODO: 를 걸어놓으면 IDE가 TODO를 따로 관리해줍니다.
나중에 로깅 컨텍스트 추가하고 나서 이벤트를 발생시키는 로직을 추가예정입니다.

vwx12345 replied:
> 아하 주석에 TODO: 를 걸어놓으면 IDE가 TODO를 따로 관리해줍니다. 나중에 로깅 컨텍스트 추가하고 나서 이벤트를 발생시키는 로직을 추가예정입니다.

제가 몰랐던 부분 배워가네요. 감사합니다.

vwx12345 commented on src/main/java/com/simpleboard/board/auth/domain/token/service/TokenDomainService.java:
```diff
@@ -0,0 +1,195 @@
+      throw new RefreshTokenInvalidException(); // TODO: 이벤트 생성
+
+    // Enroll old token to blacklist
+    blacklistRepository.save(oldClaims.tokenId(), oldClaims.expiredAt());
```
토큰 아이디와 만료시간을 저장하는데 만료시간을 저장하는 이유가 궁금합니다.
추가로 같은 로직을 가진 enrollBlacklist를 구현하셨는데 사용하지 않은 이유도 궁금합니다.

codingbaraGo replied:
만료시각은 TTL을 위해 사용합니다.
흠... 이렇게 보니 의도가 전혀 들어나지 않았군요.
도메인 계층에서 블랙리스트 엔티티를 하나 만드는게 나을까 싶네요.

enrollBlacklist의 경우, enrollBlacklist 메서드가 대체할 코드의 범위를 어디서부터로 생각하셨는지는 정확히 모르겠습니다만,
지금 선택하신 라인이 126라인이니 해당 라인을 enrollBlacklist로 대체할지에 대해 생각해보면
이미 refresh 토큰을 parsing하고, 검증까지 마친 상태입니다.
그렇기때문에 다시 토큰을 parsing하고 검증하는 로직을 반복할 필요는 없다 생각합니다.

추가로 enrollBlacklist 메서드에는 토큰 자체에 대한 유효성 검증만 존재하고, 블랙리스트 레포지토리를 조회하는 로직은 없습니다.
이 부분은 밖에서 별도로 처리해주어야하는 부분이고, 이를 위해 이미 parsing까지 완료한 TokenClaims의 상태에서는 굳이 같은 함수 내부에 있는 퍼블릭 메서드(enrollBlacklist)를 호출할 필요가 없어져버리는거죠.

vwx12345 replied:
1. TTL 부분은 제가 생각 못했네요.
2. 블랙리스트 레포지토리를 조회하는 로직을 제외하면  enrollBlacklist 와 REFRESH 토큰인지 체크하는 부분이랑 만료되었는지 확인하는 부분, 저장하는 부분까지 코드가 반복되어 리뷰 남겼습니다.

codingbaraGo replied:
블랙리스트에 레포지토리를 조회하기 위해 토큰을 파싱해서 TokenClaims를 얻어낸 뒤에, 유효성을 체크하고 블랙리스트에 조회하는 시점에서 이미 enrollBlacklist 내부에서 수행하는 로직 상당수를 수행한 상태가 됩니다.
블랙리스트 레포지토리를 조회하기 위해서는 토큰을 파싱하고 유효성 체크하는건 안할수가 없구요.
결국 enrollBlacklist를 다시 호출하고 메서드 내부에서 다시 파싱하고, 유효성 검사를 다시 진행하는건 비효율적인 흐름이 되어버립니다.

enrollBlacklist는 private이 아닌 외부 클래스용 public 메서드이니 사용 방향성이 다르다 생각해주시면 될것같습니다.

vwx12345 replied:
> 블랙리스트에 레포지토리를 조회하기 위해 토큰을 파싱해서 TokenClaims를 얻어낸 뒤에, 유효성을 체크하고 블랙리스트에 조회하는 시점에서 이미 enrollBlacklist 내부에서 수행하는 로직 상당수를 수행한 상태가 됩니다. 블랙리스트 레포지토리를 조회하기 위해서는 토큰을 파싱하고 유효성 체크하는건 안할수가 없구요. 결국 enrollBlacklist를 다시 호출하고 메서드 내부에서 다시 파싱하고, 유효성 검사를 다시 진행하는건 비효율적인 흐름이 되어버립니다.
> 
> enrollBlacklist는 private이 아닌 외부 클래스용 public 메서드이니 사용 방향성이 다르다 생각해주시면 될것같습니다.

enrollBlacklist 로 처리하면 반복되는 로직이 생길 수 밖에 없는거군요. 이해했습니다!

vwx12345 commented on src/main/java/com/simpleboard/board/auth/domain/token/service/TokenDomainService.java:
```diff
@@ -0,0 +1,195 @@
+    Role role = oldClaims.role();
+    Token accessToken = tokenProvider.issueToken(createAccessClaims(memberId, role, now));
+    TokenClaims refreshClaims = createRefreshClaims(memberId, role, oldClaims.expiredAt());
+    Token newRefreshToken = tokenProvider.issueToken(refreshClaims);
```
이부분 코드가 issueLoginToken 와 거의 유사해보이는데 함수를 분리할 생각이 있으신지 궁금합니다.

codingbaraGo replied:
새 refreshToken을 생성하는 로직에 문제가 있네요...
132라인을 
TokenClaims refreshClaims = createRefreshClaims(memberId, role, oldClaims.issueAt())
으로 바꿔야합니다.

일단 함수분리는 한번 검토를 해보겠습니다.
argument가 너무 더러워지지 않을까 싶긴 한데 잘 분리하면 코드가 깔끔해지겠네요.

codingbaraGo committed:
refactor(token): PR 리뷰 수정사항 반영
- TokenTTLPolicy 메서드 내부 if-else문 -> switch문으로 변경
- TokenDomainService 클래스 내부 코드 리팩토링 (Login 토큰 발행 로직 private 메서드로 extract)

vwx12345 reviewed (approved):
참고할 리뷰를 다 반영하셨군요. 나머지 부분은 현창님 의견이 맞는 것 같습니다. 고생하셨습니다.
